### DISCUSSION TITLE
Support for PEX (or other pre-bundled targets)

### DISCUSSION THREAD

**Discussion by dougyip** [dennisvang/tufup#116] [Q&A]
Feature Request:
Would you please consider adding a new CLI argument to 'targets' to allow it to simply copy the package file directly (and update targets.json) without needing to gzip or make patches?  We're using PEX to package our app instead of Pyinstaller so we don't need to gzip or create patches. We simply need to copy the pex from from eg. /temp/dist to /repository/targets and register it's hash/file size in target.json. This will make tufup more versatile to cater to other packaging solutions.

**Comment by dennisvang** [comment]
Hi @dougyip , thanks for your interest in `tufup`. I'll try to answer your question below.

>Would you please consider adding a new CLI argument to 'targets' to allow it to simply copy the package file directly (and update targets.json) ...

That's exactly what tufup already does. It simply copies the content of some folder into an archive, and helps you distribute this archive safely. 

The content of this folder can be anything: a simple python script; a bunch of unrelated files and folders; a PyInstaller bundle; a PEX package; and so on and so forth.

One requirement, for self-updating applications, is that the folder will also need to contain a copy of the most recent [root metadata][1] file. This file is created during the initial setup of your tufup update-repository.

>... without needing to gzip or make patches? ...

The use of gzip and patches is an implementation detail of tufup. You do not need to worry about this, as it is done automatically, and *internally*. It's just how tufup works, regardless of the type of packager that you use.

>... We're using PEX to package our app instead of Pyinstaller so we don't need to gzip or create patches. ...

Tufup does *not* require PyInstaller. In fact, tufup does not impose *any* requirement on the packaging/bundling solution.

In other words, tufup is packaging-solution-agnostic, if you like.

That means it is already possible to use tufup with PEX. Just point tufup to the folder containing your .pex files (in a subdirectory of your `./temp/dist`, probably), and make sure your pex package contains (or has access to) a copy of the root metadata file.

Note: Our [tufup-example repository][3] does use PyInstaller, but that is just because it is a very common solution, and it is what people migrating from PyUpdater are used to. As mentioned in the [tufup-example readme][2]:

>The dummy application is bundled using [PyInstaller](https://pyinstaller.org/en/stable/), but tufup works with any type of "application bundle" (i.e. just a directory with content representing the application).

[1]: https://theupdateframework.io/metadata/#root-metadata-rootjson
[2]: https://github.com/dennisvang/tufup-example#getting-started 
[3]: https://github.com/dennisvang/tufup-example

**Comment by mdhuzaifapatel** [comment]
I need your help in integrating tufup with an existing python application.
I'm using pyinstaller along with eel for bundling my app, so I was want to steps I should follow to implement auto-updates into my existing application.

In my app, once I run the pyinstaller command a  folder will be created which has all the files and an exe file. I zip this folder and distribute it to users . So please help me out how should I proceed in implementing tufup. I'm confused.

For now I'm ready with a folder which has bundled exe file with all necessary files.. 

If possible please give me a detailed explanation. 

I tried your given example and it's working fine, I have added a server's url from where the updates are getting downloaded and the example app (my_app) is updating.

Thanks a lot.

**Comment by dennisvang** [comment]
@mdhuzaifapatel see https://github.com/dennisvang/tufup-example/issues/3#issuecomment-1715725750
